java/main/com/google/jglome/Glome.java

```diff
@@ -0,0 +1,277 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.jglome;
+
+import static java.lang.System.arraycopy;
+
+import com.google.crypto.tink.subtle.X25519;
+import java.nio.ByteBuffer;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Encapsulates the logic of GLOME protocol.
+ */
+public final class Glome {
+
+  private byte[] peerKey;
+  private Mac userMacKey;
+  private Mac peerMacKey;
+  private int minPeerTagLength;
+
+  private final byte[] userPublicKey;
+  private final byte[] userPrivateKey;
+
+  public final static int MIN_CNT_VALUE = 0;
+  public final static int MAX_CNT_VALUE = 255;
+  public final static int MAX_TAG_LENGTH = 32;
+  public final static int MIN_TAG_LENGTH = 1;
+
+  /**
+   * Builder class for Glome object.
+   */
+  public static class GlomeBuilder {
+
+    private final byte[] peerKey;
+    private final int minPeerTagLength;
+    private byte[] userPublicKey;
+    private byte[] userPrivateKey;
+    private Mac userMacKey;
+    private Mac peerMacKey;
+
+    {
+      try {
+        userMacKey = Mac.getInstance("HmacSHA256");
+        peerMacKey = Mac.getInstance("HmacSHA256");
+      } catch (NoSuchAlgorithmException e) {
+        throw new AssertionError(e.getMessage());
+      }
+    }
```

Comment:
I really just wonder if the following would work (feel free to tell me "no, I tried, it doesn't"):

```java
static {
  Mac.getInstance("HmacSHA256");
}
```

If that doesn't work because you'd need a `throws`, fair enough. I don't know if there's any guidance around when to use initializer blocks (non-static) vs just doing it in the constructor though.

Reply:
I've tried and it doesn't work for several reasons. One is that I anyway need `try-catch` construction (it's not a method therefore I can't use `throws`). Talking about the place for this piece of code - I would better leave it in init block so as not to throw exception from the constructor.